bazaarvoice/emodb
# Fix inefficiency in databus poll due to unclaiming events

## Github Issue #

None

## What Are We Doing Here?

A change was introduced recently which does the following on a databus poll for _limit_ events:

- _limit_ events are polled from the event store
- The raw events are resolved in a loop in batches of 10
- If more than 100ms is spent resolving events any remaining events are unclaimed and the events resolved to that point are claimed.

After introducing this change we noticed frequent gatekeeper failures on polls. After investigation the root cause looks like it is in the way events are unclaimed.  An event is unclaimed by renewing it with a zero TTL.  However, there are two places where caching prevents unclaimed events from being immediately returned:

1. If the subscription was cached as empty then the poll will return empty for up to 1 second, even if an event was unclaimed and ready to be polled.
2. The slab cursor is cached at a point after the event in the event store.  This expires in 10 seconds, which means it could be up to 10 seconds before that unclaimed event will be returned by the event store again.

The first of the two commits in this PR addresses these issues by clearing those caches immediately when an event is renewed with a zero TTL.

The second commit in the PR is a larger change.  Instead of relying on the efficient unclaiming of events it reduces overall the need to unclaim events.  Effectively `DefaultDatabus` peek and poll operations have been changed to stream results back rather than return them as a List.  This helps with the goal of the original change, which was not to reduce the number of results returned by a poll but to reduce the time until the first poll result is made available to the API.

Since the databus peek and poll interfaces return a List and not a streaming interface the APIs were changed to return Iterators instead, although in the case of peek() a deprecated method to return List was included for backwards compatibility.  This accounts for most of the files changed, since this update affects numerous interfaces and unit tests.  

## How to Test and Verify

1. Check out this PR
2. Run through a typical battery of databus subscriptions and table updates
3. Ensure everything behaves as expected

## Risk

Because these changes affect databus polling the risk is higher than usual.  However, most of the changes in this PR involve reshuffling existing code to support streaming rather than changing any features of polling.

### Level 

`Medium

### Required Testing

`Regression`

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
